### ISSUE TITLE
Enhancement: [no-floating-promises] <Make array checking separately configurable>

### ISSUE THREAD

**Issue by jordanrastrick** [typescript-eslint/typescript-eslint#10032]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-floating-promises/

### Description

I propose that there should be configuration to disable the `no-floating-promises` rule checking for arrays specifically. The usage would look something like this:

```
    // ...
    "@typescript-eslint/no-floating-promises": ["error", { ignoreArrays: true }],
    // ...
```

There isn't much you can usefully do with a single promise other than `await` it so the lint rule rarely produces false positives in that case. However in my experience it often makes sense to construct and/or manipulate arrays of promises, only `await`ing the results via `Promise.all`, `Promise.race` etc. at a later point in the code that a human can tell leaves no promises floating but is beyond the capability of static code analysis.

### Fail

```TypeScript
const promise = new Promise((resolve, reject) => resolve('value'));
promise;

async function returnsPromise() {
  return 'value';
}
returnsPromise().then(() => {});

Promise.reject('value').catch();

Promise.reject('value').finally();
```

### Pass

```TypeScript
const promises = [1, 2, 3, 4].map(async x => x + 1);

const smalls = promises.slice(0, 2)
const bigs = promises.slice(2)

// Real world example 
export async function* merge<T>(
  iterables: AsyncIterable<T>[]
): AsyncIterable<T> {
  const iterators = iterables.map((iterable) => iterable[Symbol.asyncIterator]())
  const pulls = iterators.map(async (iterator, index) => ({
    ...(await iterator.next()),
    index,
  }))
  const indicesDone = new Set<number>()
  let stillPending = pulls
  while (stillPending.length > 0) {
    const winner = await Promise.race(stillPending)
    const { value, done, index } = winner
    if (!done) {
      pulls[index] = Promise.resolve(iterators[index].next()).then(
        ({ done, value }) => ({ done, value, index })
      )
      yield value
    } else {
      indicesDone.add(index)
    }
    // This line is currently a no-floating-promises error but wouldn't be under the new configuration:
    stillPending = pulls.filter((_, ix) => !indicesDone.has(ix))
  }
}
```

**Comment by bradzacher** [comment]
[your example doesn't currently report in the rule](https://typescript-eslint.io/play/#ts=5.5.4&fileType=.tsx&code=MYewdgzgLgBADgJxAWwJYQKYRgXhgbQEYAaGAJlIGZSAWAXQDpkBDOACmYgE8xgYAPXAD4BMANQxCASgDcAKDmhIsCCwA2a7HkQp0WBhDWpgGNgAZSZKYvDQYAI1QBzLfCRpMEA0ZNsrCgHoAmAAlDGY1GAB3EAQ1ABMYDH5mZDg1DBg5ZLhY2E4ePgAzAFdeKFRwACoYZAwEJwwAHgAVITY5GBhUKHrmewyIAC4YAEFuXgBJXoR%2BjNahfDo5KRHxwum%2Bgea2mABvTpglOx6%2BqFjXU9ntrxZ2Niu5jClhbpmn-ABlLmR7EDUGAUpu9zgg6GwpNYusdYHAShpLiCLkxWBwJnwHkiEKRUGB4skXjgRGwDl0ugwKRwoswem8zrEGGBklAIVJiIcurj8fx2V0AL6Qw4w7p44xYAAi4EyeCZURgnwwUCaYBKv3q7ShMAyKgqGgAChhRWAnLh4PDNIcogALVAZGBsaC2tQGo1OBgZY1QK0wERmF6ksnCqK4pkIU3Mam0vXuPQMWa%2BR36w3xXFOTXQ2ywPYwABuERKGFI8SlOLxyRgfNNwbAoY53SK9oAhMWmf66104Qj8FzknRTdHdJg41h-jnTI9QRBu2X%2BIwmfwWZCGF7DR0yev7dmW4Xc-nMgLXiSYNvSHm1AXS9yK%2BmyTeulxUBgErvzxg65Wn5h9u2RSmTBBJSZQF4niB4ZxvPk6yCGAWhtbAjCZbpsGAEoEAQQ0oDULgYGYGAwBAABaIo1BAZgKmNAidA8LAkjQ2IHBKWAYnheIwAAclgexMjKfEwxXPCMDlJQimcVCyMqMAhjrRNnWTVNTU7TQGBEtQZjYNgAH0cX4QkREbLkxQAqUGCtTgHh0zVIMghQB2o4cIFHUxZAUIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oDN4OBDfMwDmtYtA4BbSshTooiaBOiRwYAL4hNQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false).

Could you please clarify what cases currently error but that shouldn't?

**Comment by jordanrastrick** [comment]
The opening examples are taken from the existing Incorrect code cases given on `https://typescript-eslint.io/rules/no-floating-promises/` because I thought that was the simplest way to get simple cases of the rule that fail now but should pass. Maybe the page needs to be updated?

The latter real-world example was triggering a `no-floating-promises` error for me locally under eslint v8 but when I upgraded to eslint v9 to match the config of your playground it went away. Sorry, should have thought to check that.
